# Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Importing and Concatenating Dataframes

## Concatenating Dataframes

In [19]:
# Setting Variables
path = r'C:\Users\Saikat Halder\Documents\CareerFoundry\Data Immersion\Achievement 6\World Happiness Report Analysis 03202024'

# Initialize an empty DataFrame
df_merged = pd.DataFrame()

# Iterate through the years and merge the data
for year in range(2015, 2024):
    # Read the CSV file for the current year
    temp_df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', f'WHR_{year}.csv'))
    
    # Add a 'year' column
    temp_df['year'] = year
    
    # Concatenate the current DataFrame with the merged DataFrame
    df_merged = pd.concat([df_merged, temp_df], ignore_index=True)

# Save the merged DataFrame to a CSV file
df_merged.to_csv(os.path.join(path,'02 Data','Prepared Data','WHR_merged.csv'), index=False)


## Importing Merged Dataframe

In [23]:
# Importing Data Frame
df_combined = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','WHR_merged.csv'))

# Verifying Dataframes

In [22]:
df_combined

country                        region  happiness_score  \
0          Switzerland                Western Europe            7.587   
1              Iceland                Western Europe            7.561   
2              Denmark                Western Europe            7.527   
3               Norway                Western Europe            7.522   
4               Canada         North America and ANZ            7.427   
...                ...                           ...              ...   
1362  Congo (Kinshasa)            Sub-Saharan Africa            3.207   
1363          Zimbabwe            Sub-Saharan Africa            3.204   
1364      Sierra Leone            Sub-Saharan Africa            3.138   
1365           Lebanon  Middle East and North Africa            2.392   
1366       Afghanistan                    South Asia            1.859   

      gdp_per_capita  social_support  healthy_life_expectancy  \
0            1.39651         1.34951                  0.94143   
1            1.30232         1.40223                  0.94784   
2            1.32548         1.36058                  0.87464   
3            1.45900         1.33095                  0.88521   
4            1.32629         1.32261                  0.90563   
...              ...             ...                      ...   
1362         0.53100         0.78400                  0.10500   
1363         0.75800         0.88100                  0.06900   
1364         0.67000         0.54000                  0.09200   
1365         1.41700         0.47600                  0.39800   
1366         0.64500         0.00000                  0.08700   

      freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
0                          0.66557     0.29678                    0.41978   
1                          0.62877     0.43630                    0.14145   
2                          0.64938     0.34139                    0.48357   
3                          0.66973     0.34699                    0.36503   
4                          0.63297     0.45811                    0.32957   
...                            ...         ...                        ...   
1362                       0.37500     0.18300                    0.06800   
1363                       0.36300     0.11200                    0.11700   
1364                       0.37100     0.19300                    0.05100   
1365                       0.12300     0.06100                    0.02700   
1366                       0.00000     0.09300                    0.05900   

      year  
0     2015  
1     2015  
2     2015  
3     2015  
4     2015  
...    ...  
1362  2023  
1363  2023  
1364  2023  
1365  2023  
1366  2023  

[1367 rows x 10 columns]

In [26]:
df_combined[df_combined['country'] == 'Congo (Kinshasa)']

country              region  happiness_score  gdp_per_capita  \
119   Congo (Kinshasa)  Sub-Saharan Africa            4.517        0.000000   
282   Congo (Kinshasa)  Sub-Saharan Africa            4.272        0.056610   
440   Congo (Kinshasa)  Sub-Saharan Africa            4.280        0.092102   
601   Congo (Kinshasa)  Sub-Saharan Africa            4.245        0.069000   
752   Congo (Kinshasa)  Sub-Saharan Africa            4.418        0.094000   
912   Congo (Kinshasa)  Sub-Saharan Africa            4.311        0.062487   
1362  Congo (Kinshasa)  Sub-Saharan Africa            3.207        0.531000   

      social_support  healthy_life_expectancy  freedom_to_make_life_choices  \
119         1.001200                 0.098060                      0.226050   
282         0.806760                 0.188000                      0.156020   
440         1.229023                 0.191407                      0.235961   
601         1.136000                 0.204000                      0.312000   
752         1.125000                 0.357000                      0.269000   
912         0.833055                 0.277212                      0.364623   
1362        0.784000                 0.105000                      0.375000   

      generosity  perceptions_of_corruption  year  
119     0.248340                   0.076250  2015  
282     0.254580                   0.060750  2016  
440     0.246456                   0.060241  2017  
601     0.197000                   0.052000  2018  
752     0.212000                   0.053000  2019  
912     0.254325                   0.081467  2020  
1362    0.183000                   0.068000  2023

# Descriptive Analysis

In [28]:
# Finding the datatypes of each column
df_combined.dtypes

country                          object
region                           object
happiness_score                 float64
gdp_per_capita                  float64
social_support                  float64
healthy_life_expectancy         float64
freedom_to_make_life_choices    float64
generosity                      float64
perceptions_of_corruption       float64
year                              int64
dtype: object

In [29]:
# Number of columns and rows
df_combined.shape

(1367, 10)

In [31]:
# Min, Mean, and Max
df_combined.describe()

happiness_score  gdp_per_capita  social_support  \
count      1367.000000     1367.000000     1367.000000   
mean          5.441086        1.019422        1.045334   
std           1.117909        0.453703        0.331163   
min           1.859000        0.000000        0.000000   
25%           4.599500        0.696326        0.832022   
50%           5.440000        1.043000        1.083000   
75%           6.256300        1.338473        1.299985   
max           7.842000        2.209000        1.644000   

       healthy_life_expectancy  freedom_to_make_life_choices   generosity  \
count              1366.000000                   1367.000000  1367.000000   
mean                  0.584043                      0.450771     0.196260   
std                   0.245117                      0.156733     0.113301   
min                   0.000000                      0.000000     0.000000   
25%                   0.402301                      0.356000     0.115000   
50%                   0.612830                      0.467610     0.182700   
75%                   0.777614                      0.568685     0.252858   
max                   1.141000                      0.772000     0.838075   

       perceptions_of_corruption         year  
count                1366.000000  1367.000000  
mean                    0.132275  2018.903438  
std                     0.112555     2.561006  
min                     0.000000  2015.000000  
25%                     0.056826  2017.000000  
50%                     0.097000  2019.000000  
75%                     0.165945  2021.000000  
max                     0.587000  2023.000000

#### From the looks of it both 'healthy_life_expectancy' and 'perceptions_of_corruption' columns are missing values as the count is 1366 instead of 1367. Other than that, the minimum, maximum and mean and all within acceptable range

# Consistency Check

## Finding Mixed Type Data 

In [32]:
found_mix_type = False

for col in df_combined.columns.tolist():
  weird = (df_combined[[col]].map(type) != df_combined[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_combined[weird]) > 0:
    print (col)
    found_mix_type = True
    
if not found_mix_type:
    print('There are no mix type columns')

There are no mix type columns


## Finding and Treating Missing Values 

In [34]:
# Finding Missing Values
df_combined.isnull().sum()

country                         0
region                          0
happiness_score                 0
gdp_per_capita                  0
social_support                  0
healthy_life_expectancy         1
freedom_to_make_life_choices    0
generosity                      0
perceptions_of_corruption       1
year                            0
dtype: int64

### healthy_life_expectancy

In [37]:
# Checking the missing values of healthy_life_expectancy
df_combined[df_combined['healthy_life_expectancy'].isnull()==True]

country                        region  happiness_score  \
1328  State of Palestine  Middle East and North Africa            4.908   

      gdp_per_capita  social_support  healthy_life_expectancy  \
1328           1.144           1.309                      NaN   

      freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
1328                         0.416       0.065                      0.067   

      year  
1328  2023

In [39]:
## Checking other records for country == 'State of Palestine'
df_combined[df_combined['country'] == 'State of Palestine']

country                        region  happiness_score  \
1328  State of Palestine  Middle East and North Africa            4.908   

      gdp_per_capita  social_support  healthy_life_expectancy  \
1328           1.144           1.309                      NaN   

      freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
1328                         0.416       0.065                      0.067   

      year  
1328  2023

#### Since there is only one count of State of Palestine, we will take the mean value of the region Middle East and North Africa (2015-2023)

In [46]:
# Using lambda function to replace the missing value with the mean of the corresponding region
df_combined['healthy_life_expectancy'] = df_combined.groupby('region')['healthy_life_expectancy'].transform(lambda x: x.fillna(x.mean()))

### perceptions_of_corruption 

In [47]:
# Checking the missing values of healthy_life_expectancy
df_combined[df_combined['perceptions_of_corruption'].isnull()==True]

country                        region  happiness_score  \
489  United Arab Emirates  Middle East and North Africa            6.774   

     gdp_per_capita  social_support  healthy_life_expectancy  \
489           2.096           0.776                     0.67   

     freedom_to_make_life_choices  generosity  perceptions_of_corruption  year  
489                         0.284       0.186                        NaN  2018

In [48]:
## Checking other records for country == 'United Arab Emirates'
df_combined[df_combined['country'] == 'United Arab Emirates']

country                        region  happiness_score  \
19    United Arab Emirates  Middle East and North Africa           6.9010   
185   United Arab Emirates  Middle East and North Africa           6.5730   
335   United Arab Emirates  Middle East and North Africa           6.6480   
489   United Arab Emirates  Middle East and North Africa           6.7740   
646   United Arab Emirates  Middle East and North Africa           6.8250   
802   United Arab Emirates  Middle East and North Africa           6.7908   
959   United Arab Emirates  Middle East and North Africa           6.5610   
1107  United Arab Emirates  Middle East and North Africa           6.5760   
1255  United Arab Emirates  Middle East and North Africa           6.5710   

      gdp_per_capita  social_support  healthy_life_expectancy  \
19          1.427270        1.125750                 0.809250   
185         1.573520        0.871140                 0.729930   
335         1.626343        1.266410                 0.726798   
489         2.096000        0.776000                 0.670000   
646         1.503000        1.310000                 0.825000   
802         1.431086        1.251171                 0.787814   
959         1.555000        0.860000                 0.594000   
1107        1.998000        0.980000                 0.633000   
1255        2.015000        1.223000                 0.401000   

      freedom_to_make_life_choices  generosity  perceptions_of_corruption  \
19                        0.641570    0.264280                   0.385830   
185                       0.562150    0.265910                   0.355610   
335                       0.608345    0.360942                   0.324490   
489                       0.284000    0.186000                        NaN   
646                       0.598000    0.262000                   0.182000   
802                       0.652936    0.280656                   0.220214   
959                       0.670000    0.236000                   0.223000   
1107                      0.702000    0.204000                   0.250000   
1255                      0.745000    0.188000                   0.247000   

      year  
19    2015  
185   2016  
335   2017  
489   2018  
646   2019  
802   2020  
959   2021  
1107  2022  
1255  2023

#### Since we have records for 'United Arab Emirates' for all the years except for 2018, we will use mean grouped by Country

In [49]:
# Using lambda function to replace the missing value with the mean of the country
df_combined['perceptions_of_corruption'] = df_combined.groupby('country')['perceptions_of_corruption'].transform(lambda x: x.fillna(x.mean()))

In [51]:
# Rechecking Missing Values
df_combined.isnull().sum()

country                         0
region                          0
happiness_score                 0
gdp_per_capita                  0
social_support                  0
healthy_life_expectancy         0
freedom_to_make_life_choices    0
generosity                      0
perceptions_of_corruption       0
year                            0
dtype: int64

#### We dont have any more missing values 

# Finding Duplicate Values

In [53]:
# Finding Duplicates for df_vehicle_clean
df_combined_duplicates = df_combined[df_combined.duplicated()]
df_combined_duplicates

Empty DataFrame
Columns: [country, region, happiness_score, gdp_per_capita, social_support, healthy_life_expectancy, freedom_to_make_life_choices, generosity, perceptions_of_corruption, year]
Index: []

#### There are no duplicate values

# Exporting Cleaned Dataframe 

In [54]:
df_combined.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'WHS_merged_clean.pkl'))